### PR TITLE
Fix an issue with namespaces when an element is unmarshalled

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What**

Change the logic of namespace retrieval for element unmarshalling. The new logic:
* iterates over the element and its parents;
* creates a map of prefixes to namespaces, taking into account namespace overriding (ignore a parent namespace if a namespace with the same prefix has already been registered by a child).

Also correctly apply default namespaces (previously they would have been added like `xmlns:=value`, now they are added as `xmlns=value`).

**Why**

A valid XML lead to the following error: `expected element <Assertion> in name space urn:oasis:names:tc:SAML:2.0:assertion but have no name space`

XML:
```
<?xml version="1.0" encoding="UTF-8"?><ns3:Response xmlns:ns3="urn:oasis:names:tc:SAML:2.0:protocol" xmlns="urn:oasis:names:tc:SAML:2.0:assertion" xmlns:ns2="http://www.w3.org/2000/09/xmldsig#" xmlns:ns4="http://www.w3.org/2001/04/xmlenc#" Destination="..." ID="..." InResponseTo="..." IssueInstant="..." Version="2.0">
  <Issuer>...</Issuer>
  <ns3:Status>
    <ns3:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
  </ns3:Status>
  <Assertion ID="..." IssueInstant="...." Version="2.0">
    <Issuer>...</Issuer>
    <Signature xmlns="http://www.w3.org/2000/09/xmldsig#">
      ...
    </Signature>
     .....
  </Assertion>
</ns3:Response>
```

### CODE REVIEW

**Review by alexanderzobnin — COMMENTED** [review]
I've found a simplier solution.

```go
	for space, uri := range namespaces {
		prefix := "xmlns:" + space
		if space == "" {
			prefix = "xmlns"
		}
		doc.Root().CreateAttr(prefix, uri)
	}
```
I was confused by `NamespaceURI()` function which says that 
> If the element is part of the XML default namespace, NamespaceURI returns the empty string.

But it actually returns a default namespace. So we can skip the part inside `for _, childEl := range el.FindElements("//*") {` and just use "xmlns" key for default namespace.

**Comment by IevaVasiljeva** [comment]
> So we can skip the part inside `for _, childEl := range el.FindElements("//*") {` and just use "xmlns" key for default namespace.

That would solve the issue with default namespace not being correctly added to attributes, but it doesn't solve the problem with the namespaces not being picked up correctly. If we keep the `for _, childEl := range el.FindElements("//*") {` loop, there are two problems:
1) we are reading namespaces from all elements not only from the parent elements;
2) we are not picking up namespaces from the attributes.

**Review by alexanderzobnin — APPROVED** [review]
Ok, sounds good! Could you maybe add some comments to the namespace searching section so that will be easier to navigate and read the code in the future?

File: service_provider.go [hunk 1]
```diff
@@ -1656,18 +1656,24 @@ func findChild(parentEl *etree.Element, childNS string, childTag string) (*etree
...
 
 func elementToBytes(el *etree.Element) ([]byte, error) {
+	// Retrieve namespaces from the element itself and its parents
 	namespaces := map[string]string{}
-	for _, childEl := range el.FindElements("//*") {
-		ns := childEl.NamespaceURI()
-		if ns != "" {
-			namespaces[childEl.Space] = ns
+	currentElement := el
+	for currentElement != nil {
```
Comment (mgyongyosi) [inline service_provider.go:1662]: Could you please add a new test case for testing this functionality (elements from the default namespace)?

Comment (IevaVasiljeva) [reply service_provider.go:1662]: Added a test for it 👍

**Comment by robotdan** [comment]
Thanks!!! This is great!

**Pull request merged**